src/pools/D3MAaveDaiPool.sol

```diff
@@ -0,0 +1,203 @@
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MAaveDaiPool/not-authorized");
+        _;
+    }
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, address data);
+    event Collect(address indexed king, address[] assets, uint256 amt);
+
+    constructor(address hub_, address dai_, address pool_, address _rewardsClaimer) public {
+        pool = LendingPoolLike(pool_);
+        asset = TokenLike(dai_);
+
+        // Fetch the reserve data from Aave
+        (,,,,,,, address adai_, address stableDebt_, address variableDebt_, ,) = LendingPoolLike(pool_).getReserveData(dai_);
+        require(adai_ != address(0), "D3MAaveDaiPool/invalid-adai");
+        require(stableDebt_ != address(0), "D3MAaveDaiPool/invalid-stableDebt");
+        require(variableDebt_ != address(0), "D3MAaveDaiPool/invalid-variableDebt");
+
+        adai = ATokenLike(adai_);
+        stableDebt = ATokenLike(stableDebt_);
+        variableDebt = ATokenLike(variableDebt_);
+        rewardsClaimer = RewardsClaimerLike(_rewardsClaimer);
+
+        ATokenLike(adai_).approve(pool_, type(uint256).max);
+        TokenLike(dai_).approve(pool_, type(uint256).max);
+
+        CanLike(D3mHubLike(hub_).vat()).hope(hub_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "D3MAaveDaiPool/overflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "D3MAaveDaiPool/overflow");
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Admin ---
+    function file(bytes32 what, address data) external auth {
+        if (what == "king") king = data;
+        else revert("D3MAaveDaiPool/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // Deposits Dai to Aave in exchange for adai which gets sent to the msg.sender
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit
+    function deposit(uint256 amt) external override auth {
+        uint256 scaledPrev = adai.scaledBalanceOf(address(this));
+
+        pool.deposit(address(asset), amt, address(this), 0);
+
+        // Verify the correct amount of adai shows up
+        uint256 interestIndex = pool.getReserveNormalizedIncome(address(asset));
+        uint256 scaledAmount = _rdiv(amt, interestIndex);
```

Comment:
In limited testing during review it seems like `aDai.balanceOf()` performs in the same way as these scaled balance calculations. Is there a reason we use these extended calculations here?

@hexonaut

Reply:
Using the rebase balance results in off by one errors occasionally. If you try against a bunch of blocks you'll see.

Reply:
don't we still have that off by one with `scaledBalanceOf`? the `require` checks `>=` not `==` even with the scaling

Reply:
I believe so, but the check is there to show we have at least what we expect. If we have more then that's just a bonus. In the other case you need to over-estimate what the deficit is by just adding `+ 1` which is much less nice imo.